homeassistant/components/zha/update.py

```diff
@@ -93,127 +85,81 @@ async def async_update_data(self) -> None:
-        self._attr_latest_version = f"0x{image.version:08x}"
-        self._attr_installed_version = f"0x{current_file_version:08x}"
+    def __init__(self, entity_data: EntityData, **kwargs: Any) -> None:
+        """Initialize the ZHA siren."""
+        zha_data = get_zha_data(entity_data.device_proxy.gateway_proxy.hass)
+        assert zha_data.update_coordinator is not None
 
-        if image.metadata.changelog:
-            self._attr_release_summary = image.metadata.changelog
+        super().__init__(entity_data, coordinator=zha_data.update_coordinator, **kwargs)
+        CoordinatorEntity.__init__(self, zha_data.update_coordinator)
 
-        self.async_write_ha_state()
+    @property
+    def installed_version(self) -> str | None:
+        """Version installed and in use."""
+        return self.entity_data.entity.installed_version
 
-    @callback
-    def _update_progress(self, current: int, total: int, progress: float) -> None:
-        """Update install progress on event."""
-        # If we are not supposed to be updating, do nothing
-        if self._attr_in_progress is False:
-            return
+    @property
+    def in_progress(self) -> bool | int | None:
+        """Update installation progress.
 
-        # Remap progress to 2-100 to avoid 0 and 1
-        self._attr_in_progress = int(math.ceil(2 + 98 * progress / 100))
-        self.async_write_ha_state()
+        Needs UpdateEntityFeature.PROGRESS flag to be set for it to be used.
 
-    async def async_install(
-        self, version: str | None, backup: bool, **kwargs: Any
-    ) -> None:
-        """Install an update."""
-        assert self._latest_firmware is not None
+        Can either return a boolean (True if in progress, False if not)
+        or an integer to indicate the progress in from 0 to 100%.
+        """
+        if not self.entity_data.entity.in_progress:
+            return self.entity_data.entity.in_progress
 
-        # Set the progress to an indeterminate state
-        self._attr_in_progress = True
-        self.async_write_ha_state()
+        # Stay in an indeterminate state until we actually send something
+        if self.entity_data.entity.progress == 0:
+            return True
 
-        try:
-            result = await self.zha_device.device.update_firmware(
-                image=self._latest_firmware,
-                progress_callback=self._update_progress,
-            )
-        except Exception as ex:
-            raise HomeAssistantError(f"Update was not successful: {ex}") from ex
-
-        # If we tried to install firmware that is no longer compatible with the device,
-        # bail out
-        if result == Status.NO_IMAGE_AVAILABLE:
-            self._attr_latest_version = self._attr_installed_version
-            self.async_write_ha_state()
+        # Rescale 0-100% to 2-100% to avoid 0 and 1 colliding with None, False, and True
+        return int(math.ceil(2 + 98 * self.entity_data.entity.progress / 100))
```

Comment:
Reminder for myself: Get the proposal in for allowing a float value for the update process.